src/xwayland/xwm/settings.rs

```diff
@@ -0,0 +1,240 @@
+        Value::Color(value)
+    }
+}
+
+const fn pad(len: usize) -> usize {
+    (4 - (len % 4)) % 4
+}
+
+/// XSETTINGS Name validation error
+#[derive(Debug, thiserror::Error)]
+pub enum NameError {
+    /// Name contained an invalid character. Only a-z, A-Z, 0-9, /, _ are allowed.
+    #[error("Name contained an invalid character. Only a-z, A-Z, 0-9, /, _ are allowed.")]
+    DisallowedCharacter,
+    /// Name started or ended on a slash.
+    #[error("Name started or ended on a slash.")]
+    SlashAtStartOrEnd,
+    /// Name contained two consecutive slashes.
+    #[error("Name contained two consecutive slashes.")]
+    DoubleSlash,
+    /// Name contained a number at the start or immediately following a slash.
+    #[error("Name contained a number at the start or immediately following a slash.")]
+    LeadingNumber,
+    /// Name was empty.
+    #[error("Name was empty.")]
+    EmptyName,
+}
+
+impl XSettings {
+    pub(super) fn new(
+        conn: &RustConnection,
+        depth: u8,
+        root: Window,
+        atoms: &super::Atoms,
+    ) -> Result<XSettings, ReplyOrIdError> {
+        let win = conn.generate_id()?;
+        conn.create_window(
+            depth,
+            win,
+            root,
+            // x, y, width, height, border width
+            0,
+            0,
+            1,
+            1,
+            0,
+            WindowClass::INPUT_OUTPUT,
+            x11rb::COPY_FROM_PARENT,
+            &Default::default(),
+        )?;
+        conn.set_selection_owner(win, atoms._XSETTINGS_S0, x11rb::CURRENT_TIME)?;
+        debug!(window = win, "Created XSettings window");
+
+        let mut this = XSettings {
+            window: win,
+            atoms: *atoms,
+            values: HashMap::default(),
+            current_serial: 0,
+        };
+        this.update(conn)?;
+        Ok(this)
+    }
+
+    pub fn set(&mut self, name: impl Into<String>, value: impl Into<Value>) -> Result<(), NameError> {
+        let name = name.into();
+        if name
+            .find(|c: char| !(c.is_ascii_alphanumeric() || c == '/' || c == '_'))
+            .is_some()
+        {
+            return Err(NameError::DisallowedCharacter);
+        }
+        if name.is_empty() {
+            return Err(NameError::EmptyName);
+        }
+        if name.starts_with('/') || name.ends_with('/') {
+            return Err(NameError::SlashAtStartOrEnd);
+        }
+        if name.split('/').any(str::is_empty) {
+            return Err(NameError::DoubleSlash);
+        }
+        if name
+            .split('/')
+            .any(|s| str::starts_with(s, |c: char| c.is_ascii_digit()))
+        {
+            return Err(NameError::LeadingNumber);
+        }
+
+        self.values.insert(
+            name,
+            Setting {
+                val: value.into(),
+                last_changed_serial: self.current_serial,
+            },
+        );
+
+        Ok(())
+    }
+
+    // currently unused
```

Comment:
It doesn't matter much, but instead of commenting out working code, I'd use `#[allow(dead_code)]`.

Reply:
Fixed